odoo/addons/base/tests/test_ir_model.py

```diff
@@ -168,3 +169,112 @@ def test_create_external_id_with_space(self):
         }]
         with self.assertRaisesRegex(IntegrityError, 'ir_model_data_name_nospaces'):
             model._load_records(data_list)
+
+class TestIrModel(SavepointCase):
+    
+    @classmethod
+    def setUpClass(cls):
```

Comment:
The registry is never cleaned up.
We need to add calls to `registry.reset_changes` in both `setUpClass` and `setUp` (in `SavepointCase` or above)

Reply:
Since i'm not exactly sure of what this implies, I'm gonna leave that to you or @Elkasitu. Adding this 'blindly' breaks my test (at least the cleanup in the `setUp`) and I don't understand why:
```diff
diff --git a/odoo/tests/common.py b/odoo/tests/common.py
index 456544ec513..b08bea4960a 100644
--- a/odoo/tests/common.py
+++ b/odoo/tests/common.py
@@ -585,6 +585,12 @@ class SavepointCase(SingleTransactionCase):
     same data without influencing one another but without having to recreate
     the test data either.
     """
+
+    @classmethod
+    def setUpClass(cls):
+        super().setUpClass()
+        cls.addClassCleanup(cls.registry.reset_changes)
+
     def setUp(self):
         super().setUp()
 
@@ -594,6 +600,7 @@ class SavepointCase(SingleTransactionCase):
         self.addCleanup(envs.update, list(envs))
         self.addCleanup(envs.clear)
 
+        self.addCleanup(self.registry.reset_changes)
         self.addCleanup(self.registry.clear_caches)
         self.addCleanup(self.env.clear)
```

Reply:
@dbo-odoo I think the class cleanup of `cls.registry.reset_changes` has to be done in `SingleTransactionCase.setUpClass` right before adding `cls.registry.clear_caches` like you did for `SavepointCase.setUp`, since the data structure used for the cleanups behaves like a stack.

Also your call to `cls.addClassCleanup(cls.registry.leave_test_mode)` should be done before / right after the call to `cls.registry.enter_test_mode`, otherwise if for whatever reason your code fails, the cleanup will not be added (and thus will not be executed)

Reply:
post mortem explanation: the test mode is indeed *required* in this case, since `reset_changes` calls setup_model with a new cursor; being in test mode ensure that this is done on the same cursor.

with the test mode and the changes suggested by @Elkasitu , it works fine.